subprojects/integ-test/src/integTest/groovy/org/gradle/integtests/StaleOutputHistoryLossIntegrationTest.groovy

```diff
@@ -37,92 +39,269 @@ class StaleOutputHistoryLossIntegrationTest extends AbstractIntegrationSpec {
         mostRecentFinalReleaseExecuter.cleanup()
     }
 
-    @NotYetImplemented
     @Issue("GRADLE-1501")
-    def "production sources files are removed"() {
+    @Unroll
+    def "production sources files are removed in a single project build for #description"() {
         given:
-        def javaProjectFixture = new JavaProjectFixture()
-        def jarFile = file('build/libs/test.jar')
-        def taskPath = ':jar'
-
-        settingsFile << "rootProject.name = 'test'"
+        def javaProject = new StaleOutputJavaProject(testDirectory, null, buildDirName)
         buildFile << "apply plugin: 'java'"
 
+        if (!defaultDir) {
+            buildFile << """
+                buildDir = '$javaProject.buildDirName'
+            """
+        }
+
         when:
-        def result = runWithMostRecentFinalRelease(taskPath)
+        def result = runWithMostRecentFinalRelease(JAR_TASK_PATH)
 
         then:
-        result.executedTasks.contains(taskPath)
-        javaProjectFixture.mainClassFile.assertIsFile()
-        javaProjectFixture.redundantClassFile.assertIsFile()
-        hasDescendants(jarFile, javaProjectFixture.mainClassFile.name, javaProjectFixture.redundantClassFile.name)
+        result.executedTasks.containsAll(COMPILE_JAVA_TASK_PATH, JAR_TASK_PATH)
+        !result.output.contains(javaProject.classesDirCleanupMessage)
+        javaProject.mainClassFile.assertIsFile()
+        javaProject.redundantClassFile.assertIsFile()
+        hasDescendants(javaProject.jarFile, javaProject.mainClassFile.name, javaProject.redundantClassFile.name)
```

Comment:
We are asserting the same things over and over again. Can't we extract those to assertions on `StaleOutputJavaProject`?
We could have something linke `javaProjectFixture.assertBuiltWithRedundantFile`, `javaProjectFixture.assertBuiltWithoutRedundantFile` and `javaProjectFixture.assertCleanupExecuted`/`assertCleanupNotExecuted`. WDYT?